pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/DefaultAggregationExecutor.java

```diff
@@ -120,92 +89,96 @@ public void init() {
    * Perform aggregation on a given docIdSet.
    * Asserts that 'init' has be called before calling this method.
    *
-   * @param docIdSet
-   * @param startIndex
-   * @param length
+   * @param docIdSet block doc id set.
+   * @param startIndex start index of the block.
+   * @param length length of the block.
    */
   @Override
   public void aggregate(int[] docIdSet, int startIndex, int length) {
     Preconditions
         .checkState(_inited, "Method 'aggregate' cannot be called before 'init' for class " + getClass().getName());
 
-    fetchColumnDictIds(docIdSet, startIndex, length);
-    fetchColumnValues(startIndex, length);
+    _singleValueBlockCache.initNewBlock(docIdSet, startIndex, length);
 
-    for (int i = 0; i < _aggrFuncContextList.size(); i++) {
-      aggregateColumn(_aggrFuncContextList.get(i), _resultHolderArray[i], length);
+    for (int i = 0; i < _numAggrFunc; i++) {
+      aggregateColumn(_aggrFuncContextArray[i], _resultHolderArray[i], length);
     }
   }
 
   /**
    * Helper method to perform aggregation for a given column.
    *
-   * @param aggrFuncContext
-   * @param resultHolder
-   * @param length
+   * @param aggrFuncContext aggregation function context.
+   * @param resultHolder result holder.
+   * @param length length of the block.
    */
   private void aggregateColumn(AggregationFunctionContext aggrFuncContext, AggregationResultHolder resultHolder,
       int length) {
-    AggregationFunction function = aggrFuncContext.getAggregationFunction();
-    String aggrFuncName = aggrFuncContext.getFunctionName();
+    AggregationFunction aggregationFunction = aggrFuncContext.getAggregationFunction();
     String[] aggrColumns = aggrFuncContext.getAggregationColumns();
+    String aggrFuncName = aggregationFunction.getName();
 
     Preconditions.checkState(aggrColumns.length == 1);
     String aggrColumn = aggrColumns[0];
 
     switch (aggrFuncName) {
       case AggregationFunctionFactory.COUNT_AGGREGATION_FUNCTION:
-        function.aggregate(length, resultHolder);
-        return;
+        aggregationFunction.aggregate(length, resultHolder);
```

Comment:
do we need a interface that just takes length and resultHolder without values? Its weird all functions have to implement this method. Its probably better to have only one method aggregate(length, resultHolder, values) something like that and pass null for count aggregation function since it does not use it any way